substrate-node/pallets/pallet-tfgrid/src/node.rs

```diff
@@ -0,0 +1,211 @@
+            validate_longitude_input(&longitute_bv),
+            Self::Error::InvalidNodeLongitudeInput
+        );
+
+        Ok(Self(latitute_bv, longitute_bv, PhantomData))
+    }
+}
+
+// FIXME: did not find a way to automatically implement this.
+impl<T: Config> PartialEq for Location<T> {
+    fn eq(&self, other: &Self) -> bool {
+        self.0 == other.0 && self.1 == other.1
+    }
+}
+
+// FIXME: did not find a way to automatically implement this.
+impl<T: Config> Clone for Location<T> {
+    fn clone(&self) -> Self {
+        Self(self.0.clone(), self.1.clone(), self.2)
+    }
+}
+
+fn validate_latitude_input(input: &[u8]) -> bool {
+    match core::str::from_utf8(input) {
+        Ok(val) => {
+            if let Some(lat) = val.parse::<f32>().ok() {
+                lat >= -90.0 && lat <= 90.0
+            } else {
+                false
+            }
+        }
+        Err(_) => false,
+    }
+}
+
+fn validate_longitude_input(input: &[u8]) -> bool {
+    match core::str::from_utf8(input) {
+        Ok(val) => {
+            if let Some(long) = val.parse::<f32>().ok() {
+                long >= -180.0 && long <= 180.0
+            } else {
+                false
+            }
+        }
+        Err(_) => false,
+    }
+}
+
+/// A country name (ASCI Characters).
+#[derive(Encode, Decode, Eq, RuntimeDebug, TypeInfo, MaxEncodedLen)]
+#[scale_info(skip_type_params(T))]
+#[codec(mel_bound())]
+pub struct CountryName<T: Config>(
+    pub(crate) BoundedVec<u8, ConstU32<MAX_COUNTRY_NAME_LENGTH>>,
+    PhantomData<(T, ConstU32<MAX_COUNTRY_NAME_LENGTH>)>,
+);
+
+impl<T: Config> TryFrom<Vec<u8>> for CountryName<T> {
+    type Error = Error<T>;
+
+    /// Fallible initialization from a provided byte vector if it is below the
+    /// minimum or exceeds the maximum allowed length or contains invalid ASCII
+    /// characters.
+    fn try_from(value: Vec<u8>) -> Result<Self, Self::Error> {
+        ensure!(
+            value.len() >= MIN_COUNTRY_NAME_LENGTH.saturated_into(),
+            Self::Error::NodeCountryNameTooShort
+        );
+        let bounded_vec: BoundedVec<u8, ConstU32<MAX_COUNTRY_NAME_LENGTH>> =
+            BoundedVec::try_from(value).map_err(|_| Self::Error::NodeCountryNameTooLong)?;
+        ensure!(
+            validate_country_name(&bounded_vec),
+            Self::Error::InvalidNodeCountryName
+        );
+        Ok(Self(bounded_vec, PhantomData))
+    }
+}
+
+// FIXME: did not find a way to automatically implement this.
+impl<T: Config> PartialEq for CountryName<T> {
+    fn eq(&self, other: &Self) -> bool {
+        self.0 == other.0
+    }
+}
+
+// FIXME: did not find a way to automatically implement this.
+impl<T: Config> Clone for CountryName<T> {
+    fn clone(&self) -> Self {
+        Self(self.0.clone(), self.1)
+    }
+}
+
+fn validate_country_name(input: &[u8]) -> bool {
+    // TODO: find better alternative
+    input
+        .iter()
+        .all(|c| c.is_ascii_alphabetic() || c.is_ascii_whitespace())
```

Comment:
Maybe we should have a list of all valid "Countries" and cross check the input with the list values 

see: https://restcountries.com/v2/all

Reply:
You mean adding this countries json file to tfgrid pallet repo?
Could be good idea, this list seems good enough

Reply:
I investigated a bit and it seems not to be the best thing to do
See https://substrate.stackexchange.com/questions/2686/how-can-i-read-json-files-in-substrate

Reply:
We could extract the json data and just pour it into code, this might not be the best approach but it could work..

Reply:
Done
Very strong/restrictive validation
Would need to check in which context the create_node() extrinsic is called